src/tablestore.rs

```diff
@@ -270,15 +273,70 @@ impl TableStore {
         handle: &SSTableHandle,
         index: Arc<SsTableIndexOwned>,
         blocks: Range<usize>,
-    ) -> Result<VecDeque<Block>, SlateDBError> {
+    ) -> Result<VecDeque<Arc<Block>>, SlateDBError> {
         let path = self.path(&handle.id);
         let obj = ReadOnlyObject {
             object_store: self.object_store.clone(),
             path,
         };
-        self.sst_format
-            .read_blocks(&handle.info, index.as_ref(), blocks, &obj)
-            .await
+        let mut blocks_read = VecDeque::with_capacity(blocks.end - blocks.start);
+        let mut uncached_ranges = Vec::new();
+
+        if let Some(cache) = &self.block_cache {
+            let cached_blocks = join_all(
+                blocks
+                    .clone()
+                    .map(|block_num| async move { cache.get(&(handle.id, block_num)).await }),
+            )
+            .await;
+
+            let mut last_uncached_start = None;
+
+            for (index, block_result) in cached_blocks.into_iter().enumerate() {
+                match block_result {
+                    Some(cached_block) => {
+                        if let Some(start) = last_uncached_start.take() {
+                            uncached_ranges.push((blocks.start + start)..(blocks.start + index));
+                        }
+                        blocks_read.push_back(cached_block);
+                    }
+                    None => {
+                        last_uncached_start.get_or_insert(index);
+                    }
+                }
+            }
+
+            if let Some(start) = last_uncached_start {
+                uncached_ranges.push((blocks.start + start)..blocks.end);
+            }
+        } else {
+            uncached_ranges.push(blocks.clone());
+        }
+
+        // Read uncached blocks
+        let uncached_blocks = join_all(uncached_ranges.iter().map(|range| {
+            let obj_ref = &obj;
+            let index_ref = &index;
+            async move {
+                self.sst_format
+                    .read_blocks(&handle.info, index_ref, range.clone(), obj_ref)
+                    .await
+            }
+        }))
+        .await;
+
+        // Merge uncached blocks with blocks_read
+        for (range, range_blocks) in uncached_ranges.into_iter().zip(uncached_blocks) {
+            for (block_num, block_read) in range.zip(range_blocks?) {
+                let block = Arc::new(block_read);
+                if let Some(cache) = &self.block_cache {
```

Comment:
can we condition inserting into the cache on a parameter to this method, and pass `false` from the compactor? I'm worried the compactor is going to stomp all over the cache inserting blocks that will actually never be read again because they're being compacted away.

We can defer this to a later patch as well, we should just file an issue for it.

Reply:
Similarly, we should release blocks for SSTs that have been compacted, once the new manifest is loaded.

Reply:
Yes that make sense will update

Reply:
Yes that makes sense, will update

Reply:
> Similarly, we should release blocks for SSTs that have been compacted, once the new manifest is loaded.

Won't the blocks get pushed out of the cache based on LRU once they're deleted anyway? I don't see the harm in leaving them there. If we release the blocks we wouldn't have anything new to put in until more reads come through, right?